sematic/db/models/job.py

```diff
@@ -0,0 +1,215 @@
+        of absolute time.
+    created_at:
+        The time this record was created in the DB.
+    updated_at:
+        The time this record was last updated in the DB. See documentation in
+        last_updated_epoch_seconds for how this relates to that field.
+    """
+
+    # Q: Why duplicate data that's already in the json of
+    # status_history_serializations and value_serialization as columns?
+    # A: For two reasons:
+    #    1. It gives us freedom to refactor the dataclass for
+    #    JobStatus/ExternalJob/KubernetesExternalJob later to move fields around,
+    #    while allowing the database columns to stay stable.
+    #    2. It allows for more efficient queries on the explicit columns rather than
+    #    requiring json traversal.
+
+    __tablename__ = "jobs"
+
+    id: str = Column(types.String(), primary_key=True)
+    source_run_id: str = Column(  # type: ignore
+        types.String(),
+        nullable=False,
+    )
+    state_name: str = Column(  # type: ignore
+        types.String(),
+        nullable=False,
+    )
+
+    # Why do we need both state_name and is_active? The former is fairly
+    # free-form and subject to change. The latter should be a stable way
+    # to determine whether the job is in a terminal state or not, and
+    # whether it is expected to change.
+    is_active: int = Column(
+        types.Integer, nullable=False  # sqlite doesn't support bool
+    )
+    status_message: str = Column(types.String(), nullable=False)
+    job_type: JobType = Column(  # type: ignore
+        types.Enum(JobType),
+        nullable=False,
+    )
+    last_updated_epoch_seconds: float = Column(types.Float(), nullable=False)
+    value_serialization: Dict[str, Any] = Column(types.JSON(), nullable=False)
+    status_history_serializations: Tuple[Dict[str, Any], ...] = Column(  # type: ignore
+        types.JSON(), nullable=False
+    )
+    created_at: datetime.datetime = Column(
+        types.DateTime(), nullable=False, default=datetime.datetime.utcnow
+    )
+    updated_at: datetime.datetime = Column(
+        types.DateTime(),
+        nullable=False,
+        default=datetime.datetime.utcnow,
+        onupdate=datetime.datetime.utcnow,
+    )
+
+    @validates("job_type")
+    def validate_job_type(self, key: Any, job_type: Union[str, JobType]) -> JobType:
+        if isinstance(job_type, str):
+            return JobType[job_type]
+        elif isinstance(job_type, JobType):
+            return job_type
+        raise ValueError(f"Cannot make a JobType from {job_type}")
+
+    @classmethod
+    def from_job(cls, job: ExternalJob, run_id: str) -> "Job":
+        if not isinstance(job, ExternalJob):
+            raise ValueError(
+                f"resource must be an instance of a subclass of "
+                f"ExternalJob. Was: {job} of type "
+                f"'{type(job)}'"
+            )
+        value_serialization = value_to_json_encodable(job, type(job))
```

Comment:
We discussed this before, but I am not a big fan of using our own serializations in here.
- It is more voluminous than a simple dict
- It is more unstable (e.g. what if we make backward incompatible changes to the serialization? Do we have to update all the `Job` records?)
- It makes it harder to query (not that we are trying to at this time)

Reply:
Actually, it's *more* stable than just using regular json. Both of these changes work with Sematic dataclass serialization:

```python
@dataclass
class Foo:
    f1: int
    f2: int = 0
 
value_from_json_encodable(serialized, Foo)  # object was serialized before field f2 was added
```

and 

```python
@dataclass
class Foo:
    f1: int
 
value_from_json_encodable(serialized, Foo)  # object was serialized before field f2 was deleted
```

With regards to querying, that is one (not the only) reason why we extract the current status to columns.